File: languages/go/entity/encryption/aes_decrypter.go
package encryption

import (
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

type AesDecrypter struct {
	Alg  proto.EncryptionAlg
	Args DecrypterArgs
}

func NewAesDecrypter(args DecrypterArgs) AesDecrypter {
	return AesDecrypter{
		Alg:  proto.EncryptionAlg_A256GCM,
		Args: args,
	}
}

func (s AesDecrypter) ToProto() *proto.Decrypter {
	var localKey *proto.LocalKey
	if s.Args.LocalKey != nil {
		localKey = s.Args.LocalKey.ToProto()
	} else if s.Args.Key != "" {
		localKey = key.LocalKey{
			Key: s.Args.Key,
		}.ToProto()
	}

	var managedKey *proto.ManagedKey
	if s.Args.ManagedKey != nil {
		managedKey = s.Args.ManagedKey.ToProto()
	}

	return &proto.Decrypter{
		Alg:        s.Alg,
		LocalKey:   localKey,
		ManagedKey: managedKey,
	}
}

File: languages/js/src/entity/encryption/aes_decrypter.ts
import * as proto from "../../bridge/proto/encryption_entities";
import { KeyType, LocalKey, ManagedKey } from "../key";
import { Decrypter } from "./decrypter";
import { DecrypterArgs } from "./decrypter_args";

export class AesDecrypter implements Decrypter {
  alg: proto.EncryptionAlg;
  args: DecrypterArgs;

  constructor(key: LocalKey | ManagedKey | string) {
    this.alg = proto.EncryptionAlg.A256GCM;

    let encrypter_key;
    if (key instanceof LocalKey) {
      encrypter_key = key;
    } else if (key instanceof ManagedKey) {
      encrypter_key = key;
    } else {
      encrypter_key = new LocalKey(key, KeyType.Aes256);
    }
    this.args = new DecrypterArgs(encrypter_key);
  }

  public toProto(): proto.Decrypter {
    return proto.Decrypter.fromPartial({
      alg: this.alg,
      localKey: this.args.localKey?.toProto(),
      managedKey: this.args.managedKey?.toProto()
    });
  }
}
